geomesa-features/geomesa-feature-avro-complex/src/main/scala/org/locationtech/geomesa/features/avro/complex/AvroSchemas.scala

```diff
@@ -0,0 +1,253 @@
+package org.locationtech.geomesa.features.avro.complex
+
+import org.apache.avro.Schema.Field
+import org.apache.avro.Schema.Type._
+import org.apache.avro.io.{Decoder, Encoder}
+import org.apache.avro.{Schema, SchemaBuilder}
+import org.geotools.feature.NameImpl
+import org.geotools.gml3.GMLSchema
+import org.geotools.xs.XSSchema
+import org.locationtech.geomesa.utils.geotools.RichAttributeDescriptors._
+import org.opengis.feature.`type`._
+
+import scala.annotation.tailrec
+import scala.collection.JavaConversions._
+import scala.collection.immutable.{BitSet, Queue}
+import scala.collection.mutable
+
+object AvroSchemas {
+  val FEATURE_ID_AVRO_FIELD_NAME = "__fid__"
+  val AVRO_SIMPLE_FEATURE_VERSION = "__version__"
+  val AVRO_SIMPLE_FEATURE_USERDATA = "__userdata__"
+  val VERSION = 4
+  val AVRO_NAMESPACE = "org.geomesa"
+  val NAME_SIMPLE_CONTENT = new NameImpl(AVRO_NAMESPACE, "simpleContent")
+  val GEOT_SIMPLE_CONTENT = new NameImpl("simpleContent")
+  val USERDATAITEM_SCHEMA = SchemaBuilder.record("userDataItem")
+    .namespace(AVRO_NAMESPACE)
+    .fields()
+    .name("class").`type`.stringType().noDefault()
+    .name("key").`type`.stringType().noDefault()
+    .name("value").`type`.stringType().noDefault().endRecord()
+
+  trait FieldNameEncoder {
+    def encode(name:String): String
+    def decode(name:String): String
+
+    def decode(s:Schema): Name = {
+      s.getType match {
+        case RECORD|ENUM|FIXED => new NameImpl(decode(s.getNamespace), decode(s.getName))
+        case _ => new NameImpl(decode(s.getProp("namespace")), decode(s.getName))
+      }
+    }
+    def decode(f:Field, parent:Schema): Name = {
+      val ns = Option(f.getProp("namespace")).getOrElse(parent.getNamespace)
+      new NameImpl(decode(ns), decode(f.name))
+    }
+  }
+
+  object FieldNameEncoderV1 extends FieldNameEncoder{
+    private val radix = "0123456789ABCDEF"
+    private val allowedCharsInit:BitSet = ('A'.toInt.to('Z')++'a'.toInt.to('z')).map(identity)(scala.collection.breakOut)

```

Comment:
Also, since we review code on GitHub, we tend to wrap before 100 (or maybe 120) characters so that lines are completely visible on the screen.

This is despite everyone having 24+ inch monitors;).